src/relations/db.py

```diff
@@ -134,6 +134,12 @@ def _on_relation_changed(self, event: RelationChangedEvent) -> None:
             # created in a previous relation changed event.
             user = f"relation-{event.relation.id}"
             password = unit_relation_databag.get("password", new_password())
+
+            # Store the user, password and database name in the secret store to be accessible by
+            # non-leader units when the cluster topology changes.
+            self.charm.set_secret("app", user, password)
+            self.charm.set_secret("app", f"{user}-database", database)
```

Comment:
I thought secrets are available for Juju 3.0+, TIL.

Reply:
> I thought secrets are available for Juju 3.0+, TIL.

Those are convenience wrapper methods within the charms, that currently just put things into databags. In the future they will handle whether to use secrets for juju 3 or databags for juju 2